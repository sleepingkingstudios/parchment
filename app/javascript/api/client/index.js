import {
  applyMiddleware,
  selectMiddleware,
} from 'api/middleware/utils';
import {
  valueOrDefault,
} from 'utils/object';

import generateActions from './actions';
import generateHooks from './hooks';
import generateInitialState from './initialState';
import generateReducer from './reducer';
import generateRequest from './request';
import generateSelector from './selector';

const wrapReducer = ({ middleware, reducer }) => applyMiddleware(
  reducer,
  selectMiddleware(middleware, 'handleAction'),
);

const buildClient = (options) => {
  const {
    namespace,
    url,
  } = options;
  const data = valueOrDefault(options.data, {});
  const method = valueOrDefault(options.method, 'GET');
  const middleware = valueOrDefault(options.middleware, []);

  const actions = generateActions({ namespace });
  const initialState = generateInitialState({ data });
  const performRequest = generateRequest({
    actions,
    method,
    middleware,
    namespace,
    url,
  });
  const hooks = generateHooks({ performRequest });
  const reducer = wrapReducer({
    middleware,
    reducer: generateReducer({ actions, initialState }),
  });
  const selector = generateSelector({ namespace });

  return {
    actions,
    hooks,
    method,
    middleware,
    namespace,
    performRequest,
    reducer,
    selector,
    url,
  };
};

export default buildClient;
