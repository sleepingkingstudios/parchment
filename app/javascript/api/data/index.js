export { default as generateHooks } from './hooks';

export { default as generateReducer } from './reducer';
