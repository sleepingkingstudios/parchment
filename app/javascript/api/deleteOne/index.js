import ApiEndpoint from '../endpoint/index';
import DeleteOneRequest from './request';
import generateActions from '../endpoint/actions';
import generateInitialState from '../endpoint/initialState';
import generateHooks from './hooks';
import generateReducer from './reducer';

class DeleteOneEndpoint extends ApiEndpoint {
  constructor(options) {
    super(
      Object.assign(
        {
          generateActions,
          generateHooks,
          generateInitialState,
          generateReducer,
          generateRequest: opts => new DeleteOneRequest(opts),
        },
        options,
      ),
    );
  }
}

export default DeleteOneEndpoint;
