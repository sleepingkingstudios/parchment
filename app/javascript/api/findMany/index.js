import ApiEndpoint from '../endpoint/index';
import FindOneRequest from './request';
import generateActions from '../endpoint/actions';
import generateInitialState from '../endpoint/initialState';
import generateReducer from './reducer';

class FindManyEndpoint extends ApiEndpoint {
  constructor(options) {
    super(
      Object.assign(
        {
          generateActions,
          generateInitialState,
          generateReducer,
          generateRequest: opts => new FindOneRequest(opts),
        },
        options,
      ),
    );
  }
}

export default FindManyEndpoint;
