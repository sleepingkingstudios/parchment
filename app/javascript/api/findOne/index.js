import generateActions from '../endpoint/actions';
import generateInitialState from '../endpoint/initialState';
import generateReducer from './reducer';
import FindOneRequest from './request';

import {
  applyMiddleware,
  selectMiddleware,
} from '../middleware/utils';

const generateFindOneRequest = ({
  data,
  middleware,
  namespace,
  url,
}) => {
  const actions = generateActions({ namespace });
  const request = new FindOneRequest({ actions, namespace, url });
  const initialState = generateInitialState({ data, namespace });
  const reducer = applyMiddleware(
    generateReducer({ actions, initialState }),
    selectMiddleware(middleware, 'handleAction'),
  );

  request.handleFailure = applyMiddleware(
    request.handleFailure,
    selectMiddleware(middleware, 'handleFailure'),
  );
  request.handleSuccess = applyMiddleware(
    request.handleSuccess,
    selectMiddleware(middleware, 'handleSuccess'),
  );

  return {
    actions,
    namespace,
    reducer,
    request,
  };
};

export default generateFindOneRequest;
