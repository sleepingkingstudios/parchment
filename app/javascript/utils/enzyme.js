import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
