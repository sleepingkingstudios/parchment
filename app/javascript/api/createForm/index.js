import pluralize from 'pluralize';

import FormEndpoint from '../form';
import authorization from '../middleware/authorization';
import generateAlerts from '../middleware/alerts';
import generateRedirectToShow from '../middleware/redirectToShow';
import {
  exists,
  valueOrDefault,
} from '../../utils/object';
import { capitalize } from '../../utils/string';

const generateNamespace = (options) => {
  if (exists(options.namespace)) { return options.namespace; }

  const { resourceName } = options;

  return `${pluralize(resourceName)}/create${capitalize(resourceName)}Form`;
};

const generateUrl = (options) => {
  const { namespace } = options;

  if (exists(options.url)) { return options.url; }

  return `/api/${namespace.replace(/\/\w+$/, '')}`;
};

const createFormEndpoint = (options) => {
  const { resourceName } = options;
  const alerts = generateAlerts({
    action: 'create',
    resourceName,
    pending: true,
    failure: true,
    success: true,
  });
  const redirect = generateRedirectToShow({
    resourceName,
    on: 'success',
  });
  const data = valueOrDefault(options.data, {});
  const middleware = [
    authorization,
    alerts,
    redirect,
  ];
  const namespace = generateNamespace(options);
  const url = generateUrl({ ...options, namespace });
  const endpoint = new FormEndpoint({
    data,
    middleware,
    namespace,
    url,
  });

  endpoint.middleware = middleware;
  endpoint.options = options;
  endpoint.type = 'api/createForm';

  return endpoint;
};

export default createFormEndpoint;
