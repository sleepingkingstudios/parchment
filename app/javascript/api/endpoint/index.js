import defaultGenerateHooks from './hooks';
import defaultGenerateSelector from './selector';
import {
  applyMiddleware,
  selectMiddleware,
} from '../middleware/utils';
import { valueOrDefault } from '../../utils/object';

const applyMiddlewareToRequest = (request, middleware) => {
  request.handleFailure = applyMiddleware(
    request.handleFailure,
    selectMiddleware(middleware, 'handleFailure'),
  );
  request.handlePending = applyMiddleware(
    request.handlePending,
    selectMiddleware(middleware, 'handlePending'),
  );
  request.handleSuccess = applyMiddleware(
    request.handleSuccess,
    selectMiddleware(middleware, 'handleSuccess'),
  );

  return request;
};

class ApiEndpoint {
  constructor(options) {
    const {
      generateActions,
      generateHooks,
      generateInitialState,
      generateReducer,
      generateRequest,
      generateSelector,
      method,
      middleware,
      namespace,
      url,
    } = options;
    const actions = generateActions({ namespace });
    const data = valueOrDefault(options.data, {});
    const initialState = generateInitialState({ data });
    const reducer = generateReducer({ actions, initialState });
    const request = generateRequest({
      actions,
      method,
      namespace,
      url,
    });
    const { performRequest } = request;
    const selector = valueOrDefault(generateSelector, defaultGenerateSelector)({ namespace });
    const hooks = valueOrDefault(generateHooks, defaultGenerateHooks)({
      actions,
      performRequest,
      selector,
    });

    this.actions = actions;
    this.hooks = hooks;
    this.namespace = namespace;
    this.reducer = applyMiddleware(
      reducer,
      selectMiddleware(middleware, 'handleAction'),
    );
    this.request = applyMiddlewareToRequest(request, middleware);
    this.selector = selector;
  }
}

export default ApiEndpoint;
