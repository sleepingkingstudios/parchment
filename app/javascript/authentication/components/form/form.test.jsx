import React from 'react';
import { shallow } from 'enzyme';

import LoginForm from './form';
import { INITIALIZED } from '../../../api/status';
import {
  toHaveFormControl,
  toHaveFormGroup,
  toHaveFormInput,
} from '../../../utils/enzyme';

expect.extend({
  toHaveFormControl: toHaveFormControl(expect),
  toHaveFormGroup: toHaveFormGroup(expect),
  toHaveFormInput: toHaveFormInput(expect),
});

describe('<AuthenticationLoginForm />', () => {
  const onChangeAction = jest.fn(
    ({ propName, value }) => ({ payload: { propName, value } }),
  );
  const onSubmitAction = jest.fn(() => ({ ok: true }));
  const data = { username: 'Alan Bradley', password: 'tronlives' };
  const errors = {};
  const defaultProps = {
    data,
    errors,
    onChangeAction,
    onSubmitAction,
    status: INITIALIZED,
  };
  const form = {
    data,
    errors,
    path: [],
    onChangeAction,
    onSubmitAction,
  };

  it('should render a form', () => {
    const rendered = shallow(<LoginForm {...defaultProps} />);

    expect(rendered).toHaveDisplayName('Form');
    expect(rendered).toHaveProp({ form });
  });

  it('should render the password field', () => {
    const rendered = shallow(<LoginForm {...defaultProps} />);

    expect(rendered).toHaveFormGroup('PasswordField', (formGroup) => {
      expect(formGroup).toHaveFormInput({
        inputType: 'FormPasswordInput',
        form,
        prop: 'password',
      });
    });
  });

  it('should render the username field', () => {
    const rendered = shallow(<LoginForm {...defaultProps} />);

    expect(rendered).toHaveFormGroup('UsernameField', (formGroup) => {
      expect(formGroup).toHaveFormInput({
        inputType: 'FormInput',
        form,
        prop: 'username',
      });
    });
  });

  it('should render the submit button', () => {
    const rendered = shallow(<LoginForm {...defaultProps} />);
    const { status } = defaultProps;

    expect(rendered).toHaveFormGroup('SubmitButton', (formGroup) => {
      expect(formGroup).toHaveFormControl('FormSubmitButton', {
        children: 'Log In',
        form,
        status,
      });
    });
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<LoginForm {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });
});
