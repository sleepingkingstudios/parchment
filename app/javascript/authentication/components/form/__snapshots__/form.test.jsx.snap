// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthenticationLoginForm /> should match the snapshot 1`] = `
<Form
  className="login-form"
  form={
    Object {
      "data": Object {
        "password": "tronlives",
        "username": "Alan Bradley",
      },
      "errors": Object {},
      "onChangeAction": [MockFunction],
      "onSubmitAction": [MockFunction],
      "path": Array [],
    }
  }
>
  <FormRow
    align="left"
    className=""
    verticalAlign="top"
  >
    <UsernameField
      colWidth={6}
      form={
        Object {
          "data": Object {
            "password": "tronlives",
            "username": "Alan Bradley",
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [],
        }
      }
    />
    <PasswordField
      colWidth={6}
      form={
        Object {
          "data": Object {
            "password": "tronlives",
            "username": "Alan Bradley",
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [],
        }
      }
    />
  </FormRow>
  <FormRow
    align="right"
    className=""
    verticalAlign="top"
  >
    <SubmitButton
      colWidth="3"
      form={
        Object {
          "data": Object {
            "password": "tronlives",
            "username": "Alan Bradley",
          },
          "errors": Object {},
          "onChangeAction": [MockFunction],
          "onSubmitAction": [MockFunction],
          "path": Array [],
        }
      }
      status="initialized"
    />
  </FormRow>
</Form>
`;
